### PR TITLE
Delay reading from StorageKafka to allow multiple reads in materialized views

### LINKED ISSUE #58405: Materialized View Regression for Kafka input tables (23.12.1 version)
**Describe what's wrong**

We are using Kafka Engine table to ingest the data from Kafka to Clickhouse. Then we connect materialized view to it and store the results to target table. As we want to achieve first write wins strategy we need to check the target contents first to deduplicate incoming records. The fastest way based on benchmark was to use IN operator (JOINS were slow). To be able to do that we need to read records from source table (insert batch) twice inside materialized view. AFAIK it should be supported.

This stops to work in 23.12 with this error:

```
com.clickhouse.client.ClickHouseException: Code: 620. DB::Exception: Direct select is not allowed. To enable use setting 
`stream_like_engine_allow_direct_select`: While processing id IN ((SELECT id FROM kafka_input GROUP BY id) AS _subquery72): 
While processing id NOT IN ((SELECT id FROM deduplicate WHERE id IN (SELECT id FROM kafka_input GROUP BY id)) AS 
_subquery71). (QUERY_NOT_ALLOWED) (version 23.12.1.1368 (official build))
```

The view looks like this. Notice that we query FROM kafka_input twice (source table for materialized view)
```
CREATE MATERIALIZED VIEW IF NOT EXISTS deduplicate_mv TO deduplicate 
AS SELECT 
	id,time,any(value) AS value
FROM kafka_input 
WHERE id NOT IN (
	SELECT id FROM deduplicate WHERE id IN (
		SELECT id FROM kafka_input GROUP BY id)
	)
GROUP BY id,time;
```

Similar example using NULL engine works as expected. Here is a fiddle:
https://fiddle.clickhouse.com/a0207085-22c8-44f4-9c7f-acf751058644

**Does it reproduce on recent release?**

Yes. It is a regression of 23.12 release.

**How to reproduce**

* Use ClickHouse 23.12
* Use Kafka engine input table
* Use Materialized view that uses Kafka input table twice as in our example

**Expected behavior**

It should be possible to create such a materialized view as query is not querying kafka table itself, but the batch of inserts produced by Kafka engine table.

### PR DESCRIPTION
**Pull request by antaljanosbenjamin**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fix reading multiple times from KafkaEngine in materialized views.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

Closes #58405.

### CODE REVIEW

**Comment by robot-ch-test-poll** [comment]
*This is an automated comment for commit 81b5d8fddddbac7c59e3a15855a0d2cf842f26b1 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/58477/81b5d8fddddbac7c59e3a15855a0d2cf842f26b1/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/58477/81b5d8fddddbac7c59e3a15855a0d2cf842f26b1/ci_running.html">⏳ pending</a></td></tr>
<tbody>
</table>

**Comment by Avogar** [comment]
Can we add a test?

**Comment by antaljanosbenjamin** [comment]
Yes, I was planning to add a test, but first wanted to see that I haven't broke anything else.

File: src/Storages/Kafka/StorageKafka.cpp [hunk 1]
```diff
@@ -174,6 +178,92 @@ struct StorageKafkaInterceptors
...
+        : ISourceStep{DataStream{.header = storage_snapshot_->getSampleBlockForColumns(column_names_)}}
+        , column_names{column_names_}
+        , storage{storage_}
+        , storage_snapshot{storage_snapshot_}
+        , storage_limits{query_info.storage_limits}
+        , context{context_}
+    {
+    }
+
+    String getName() const override { return "ReadFromStorageKafka"; }
```
Comment (KochetovNicolai) [inline src/Storages/Kafka/StorageKafka.cpp:199]: Minor, but usually I use `ReadStorageKafka` for both class name and getName()

**Review by KochetovNicolai — APPROVED** [review]

**Comment by antaljanosbenjamin** [comment]
Upgrade check: #57893

File: src/Storages/Kafka/StorageKafka.cpp (hunk 1 above)
Comment (antaljanosbenjamin) [reply src/Storages/Kafka/StorageKafka.cpp:199]: I will fix this in my next PR where I make the same fix for `StorageFileLog`. Is that okay?

**Comment by antaljanosbenjamin** [comment]
The stateless check is flaky, so I have a strong assumption that it is not connected. I will merge this PR now, but will check the tests later.

**Comment by antaljanosbenjamin** [comment]
https://github.com/ClickHouse/ClickHouse/pull/58809 for the flaky `00993_system_parts_race_condition_drop_zookeeper` test.

**Comment by antaljanosbenjamin** [comment]
Both the upgrade and stateless check is finished successfully after rerun so I will manually set the mergeable check and merge this pr. 

https://github.com/ClickHouse/ClickHouse/actions/runs/7408373875/job/20488640555?pr=58477
https://github.com/ClickHouse/ClickHouse/actions/runs/7408373875/job/20488658938?pr=58477

**Pull request merged**